### DISCUSSION TITLE
Repeated state node logic

### DISCUSSION THREAD

**Discussion by claidler** [statelyai/xstate#3283] [Q&A] [answered]
Would really appreciate some help on this one please... 

I'm just wondering what would be the best way to write duplicated state node logic, as in the example below?

```typescript
const exampleMachine = createMachine({
    id: 'example',
    initial: 'stateNodeA',

    states: {
        stateNodeA: {
            blue: {},
            purple: {},
            orange: {}
            // lots more states same as stateNodeC
        },
        stateNodeB: {},
        stateNodeC: {
            blue: {},
            purple: {},
            orange: {}
            // lots more states same as stateNodeA
        }
    }
});
```

Would it be considered good practice to do something like the following?

```typescript

const stateNodeAC = {
      blue: {},
      purple: {},
      orange: {}
      // lots more states that stateNodeA and stateNodeC have in common
}

const exampleMachine = createMachine({
    id: 'example',
    initial: 'stateNodeA',

    states: {
        stateNodeA: {
            ...stateNodeAC
        },
        stateNodeB: {},
        stateNodeC: {
            ...stateNodeAC
        }
    }
});
```

**Comment by Andarist** [comment, marked as answer]
The proposed solution works at runtime but it hinders our static analysis/transformation possibilities so I don't truly recommend it. The question though is - what's your use case for this? Can't you group those states to be a single state? And maybe add a context variable~ to distinguish before "variations" of this single state? How sure are you that those 2 states would stay truly identical in the future? Note that, for example, you couldn't use IDs within such a duplicated state as IDs should be unique across the whole config. It's just too easy to make some accidental mistake with such a spread.

**Reply by claidler** [reply]
Thanks for your time on this one @Andarist 
Yeah that makes a lot of sense. I guess technically those inner states could change based on the outer states, in which case I should keep the inner states separate, even though there is a lot of code repetition?
If there was 100% no chance on this earth those inner states could change based on the outer state then I should treat them as one state.
It's a work related thing, so I'm not sure how much I'm allowed to discuss on a public forum, so sorry I can't give the direct use case, but as an example..

I have the state of a dog, human and car.

Dog and human will have a lot of parallel states in common: breathing, walking, sleeping, eating. My state machine might not include states that exist on only dog or human, such as barking or reading. However, because they technically could have different inner states I should duplicate the common inner states no matter how many there are? 
I suppose in the future walking could then include some details about walking on 4 legs vs 2 legs, obviously we walk differently from dogs, so the walking state isn't exactly the same state.
I'm just struggling with knowing what is an allowed amount of duplication in a state machine. Should we throw the concept of DRY out of the window? Some duplication is okay right?

> maybe add a context variable~ to distinguish before "variations" of this single state

On this, I've been a bit cautious about defining anything other than infinite state in context, but would you say it's okay in certain situations to use finite state in context? I know you don't set any hard rules, but I want to be doing this the right way. Really appreciate the help

**Reply by Andarist** [reply]
> I'm just struggling with knowing what is an allowed amount of duplication in a state machine.

It's like asking what is an allowed amount of duplication in **code**. There is no real answer to that as it always depends on the code/use case itself.

> Should we throw the concept of DRY out of the window?

It also depends on what you mean by DRY. You can apply this rule in a really hardcore way - but there are also WET and AHA principles out there. It's all a balance and should be applied wisely. 

Patterns to deduplicate logic in XState machines are definitely under exploration and we could have more resources about that. In practice, in my projects, I didn't find as many problems in which I would like to actually share the very same logic between machines. And when I did I extracted the shared logic to another machine that could be invoked/spawned by the parent and then used synchronization (through events) to share the information I needed between that child machine and its parent. YMMV

> Some duplication is okay right?

Yes, some duplication is OK (at least, that's my opinion :p)

> On this, I've been a bit cautious about defining anything other than infinite state in context, but would you say it's okay in certain situations to use finite state in context?

Yes, I would say that it can be OK.

> I know you don't set any hard rules, but I want to be doing this the right way.

In the same way, as there are no hard rules like this here, there is no right way to approach this 😉 There are as many ways to model software as there are authors of the given software. If you manage to solve your problem - I encourage you to not be bothered with "right ways" of doing X (unless you see flaws in your approach but then you probably have recognized that you didn't solve your problem in full).

**Reply by claidler** [reply]
This has been really helpful, thanks again! It all makes sense.
